core/unit_test/tools/include/ToolTestingUtilities.hpp

```diff
@@ -0,0 +1,1175 @@
+        [](const uint32_t num_settings,
+           Kokkos::Tools::Experimental::ToolSettings* settings) {
+          found_events.push_back(std::make_shared<RequestToolSettingsEvent>(
+              num_settings, *settings));
+        });
+  }
+}
+template <int priority>
+void listen_tool_events_impl(std::integral_constant<int, priority>,
+                             ToolValidatorConfiguration&) {}
+
+template <class Config>
+void invoke_config(ToolValidatorConfiguration& in, Config conf,
+                   std::true_type) {
+  conf(in);
+}
+template <class Config>
+void invoke_config(ToolValidatorConfiguration&, Config, std::false_type) {}
+
+template <int priority, class Config, class... Configs>
+void listen_tool_events_impl(std::integral_constant<int, priority> prio,
+                             ToolValidatorConfiguration& in, Config conf,
+                             Configs... configs) {
+  invoke_config(in, conf,
+                std::integral_constant<bool, priority == conf.value>{});
+  listen_tool_events_impl(prio, in, configs...);
+}
+template <class... Configs>
+void listen_tool_events(Configs... confs) {
+  ToolValidatorConfiguration conf;
+  listen_tool_events_impl(std::integral_constant<int, 0>{}, conf, confs...);
+  listen_tool_events_impl(std::integral_constant<int, 1>{}, conf, confs...);
+  listen_tool_events_impl(std::integral_constant<int, 2>{}, conf, confs...);
+  set_tool_events_impl(conf);
+}
+
+/**
+ * @brief This is the main entry point people will use to test their programs
+ *        Given a lambda representing a code region, and a set of matchers on
+ * tools events, verify that the given lambda produces events that match those
+ * expected by the matchers
+ *
+ * @tparam Lambda Type of lam
+ * @tparam Matchers Type of matchers
+ * @param lam The code region that will produce events
+ * @param matchers Matchers for those events, lambdas that expect events and
+ * compare them
+ * @return true if all events are consumed, all matchers are invoked, and all
+ * matchers success, false otherwise
+ */
+template <class Lambda, class... Matchers>
+bool validate_event_set(const Lambda& lam, const Matchers... matchers) {
+  // First, erase events from previous invocations
+  found_events.clear();
+  // Invoke the lambda (this will populate found_events, via tooling)
+  lam();
+  // compare the found events against the expected ones
+  auto success = compare_event_vectors(found_events, matchers...);
+  if (!success) {
+    // on failure, print out the events we found
+    for (const auto& event : found_events) {
+      std::cout << event->repr() << std::endl;
+    }
+  }
+  return success;
+}
+/**
+ * @brief Analogous to validate_event_set up above, except rather than
+ *        comparing to matchers, this just returns the found event vector
+ *
+ * @tparam Lambda as in validate_event_set
+ * @param lam as in validate_event_set
+ * @return auto
+ */
+template <class Lambda>
+auto get_event_set(const Lambda& lam) {
```

Comment:
This is not used at all at the moment AFAICT.

Reply:
Not yet. The idea is to enable folks who want to do something custom, to do that thing. I'll add a test for it though.

Reply:
Added a test